tests/e2e/e2e_setup_test.go

```diff
@@ -325,3 +336,85 @@ func noRestart(config *docker.HostConfig) {
 		Name: "no",
 	}
 }
+
+func (s *IntegrationTestSuite) initUpgrade() {
+	// submit, deposit, and vote for upgrade proposal
+	s.submitProposal(s.chains[0])
+	s.submitProposal(s.chains[1])
+	s.depositProposal(s.chains[0])
+	s.depositProposal(s.chains[1])
+	s.voteProposal(s.chains[0])
+	s.voteProposal(s.chains[1])
+
+	// wait till all chains halt at upgrade height
+	for _, chain := range s.chains {
+		for i := range chain.Validators {
+			s.T().Logf("waiting to reach upgrade height for chain-id: %s", chain.ChainMeta.Id)
+			s.Require().Eventually(
+				func() bool {
+					out := s.chainStatus(chain, i)
```

Comment:
Passing an index to a function like this is a code smell. It basically couples the function on its parameter. I'd like to suggest passing `s.valResources[c.ChainMeta.Id][i].Container.ID` here directly

Reply:
Didn't know this but it makes total sense, thanks for explaining the reasoning behind it!